File: R/quick_sort.R
# Partition function to rearrange the array around a pivot
partition <- function(arr, low, high) {
  pivot <- arr[high]  # Choose the last element as pivot
  i <- low - 1        # Index of smaller element
  
  for (j in low:(high - 1)) {
    if (arr[j] <= pivot) {
      i <- i + 1
      # Swap elements around pivot
      temp <- arr[i]
      arr[i] <- arr[j]
      arr[j] <- temp
    }
  }
  # Place pivot element in the correct position
  temp <- arr[i + 1]
  arr[i + 1] <- arr[high]
  arr[high] <- temp
  return(i + 1)
}

# Quick Sort function using recursion
quick_sort <- function(arr, low, high) {
  if (low < high) {
    pivot_index <- partition(arr, low, high)
    
    # Recursively sort elements before and after partition
    arr <- quick_sort(arr, low, pivot_index - 1)
    arr <- quick_sort(arr, pivot_index + 1, high)
  }
  return(arr)
}

# Example usage
array <- c(10, 7, 8, 9, 1, 5)
cat("Unsorted array:", array, "\n")
sorted_array <- quick_sort(array, 1, length(array))
cat("Sorted array:", sorted_array, "\n")

File: Java/QuickSort.java
public class QuickSort {

    // Function to partition the array around a pivot element
    public static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];  // Choose the last element as pivot
        int i = low - 1;        // Index of smaller element

        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                int temp = arr[i];  // Swap elements around pivot
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    // Quick Sort function using recursion
    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pivotIndex = partition(arr, low, high);

            // Recursively sort elements before and after partition
            quickSort(arr, low, pivotIndex - 1);
            quickSort(arr, pivotIndex + 1, high);
        }
    }

    public static void main(String[] args) {
        int[] arr = {10, 7, 8, 9, 1, 5};
        
        System.out.print("Unsorted array: ");
        for (int num : arr) System.out.print(num + " ");
        System.out.println();

        quickSort(arr, 0, arr.length - 1);

        System.out.print("Sorted array: ");
        for (int num : arr) System.out.print(num + " ");
    }
}
